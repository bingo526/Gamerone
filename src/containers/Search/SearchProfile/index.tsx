import React, {
  useState,
  useEffect,
  useCallback,
  useRef,
  useLayoutEffect,
} from 'react';
import { useHistory } from 'react-router-dom';
import DebouncedInput from 'components/common/DebouncedInput';

import ProfileListItem from '../ProfileListItem';
import useSearchApi from 'lib/useSearchApi';

// api
import * as ProfileApi from 'api/profile';

import './style.scss';
import Error from 'components/common/Message/Error';
import { User } from 'interfaces';

export interface SerachProfileProps {
  navigate?: boolean;
  multiple?: boolean;
  clearInput?: boolean;
  initLoad?: boolean;
  onAdd?: (id: number) => void;
  onRemove?: (id: number) => void;
  limit?: number;
  blockedIds?: number[];
}

function SearchProfile({
  navigate = true,
  multiple = false,
  clearInput = true,
  initLoad = false,
  onAdd,
  onRemove,
  limit,
  blockedIds = [],
  ...rest
}: SerachProfileProps & React.HTMLProps<HTMLDivElement>) {
  const searchRef = useRef<HTMLInputElement>(null);
  const dropdownRef = useRef<HTMLDivElement>(null);

  const history = useHistory();
  const [query, setQuery] = useState<string | null | undefined>(
    initLoad ? '' : null,
  );
  const [dropdownHidden, setDropdownHidden] = useState(!initLoad);
  const [{ results: profiles, isSearching, isError }, doFetch] = useSearchApi<
    User
  >(ProfileApi.searchProfile, initLoad ? '' : null, []);
  const [selectedProfiles, setSelectedProfiles] = useState<User[]>([]);

  useEffect(() => {
    doFetch(query);
    if (query === null) setQuery(undefined);
  }, [query, doFetch]);

  useLayoutEffect(() => {
    if (searchRef.current) searchRef.current.focus();
  }, [searchRef]);

  const handleInputChange = useCallback(
    (e) => {
      setQuery(e.target.value);
    },
    [setQuery],
  );

  const handleEscKey = useCallback((e) => {
    if (e.keyCode === 27) {
      setDropdownHidden(true);
      if (searchRef.current) searchRef.current.blur();
    }
  }, []);

  const handleDocumentMousedown = useCallback((e) => {
    if (
      dropdownRef.current &&
      !dropdownRef.current.contains(e.target) &&
      searchRef.current &&
      searchRef.current !== e.target
    ) {
      setDropdownHidden(true);
    }
  }, []);

  const handleInputFocus = useCallback(() => {
    setDropdownHidden(false);
  }, []);

  const handleItemSelect = useCallback(
    (profile: User) => {
      if (!navigate) {
        if (onAdd) onAdd(profile.id);

        // unless it surpasses the limit
        if (!limit || selectedProfiles.length < limit)
          setSelectedProfiles([...selectedProfiles, profile]);

        // selecting one-item list should hide the dropdown
        if (profiles.length === 1 && profiles[0].id === profile.id)
          setDropdownHidden(true);
      } else if (navigate) {
        history.push('/' + profile.username);
      }

      if (clearInput) {
        setQuery(null);
        setDropdownHidden(true);
        if (searchRef.current) searchRef.current.blur();
      } else if (multiple) {
        if (searchRef.current) searchRef.current.focus();
      } else {
        setDropdownHidden(true);
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
      profiles,
      selectedProfiles,
      navigate,
      history,
      multiple,
      clearInput,
      limit,
    ],
  );

  const handleItemUnselect = useCallback(
    (id: number) => {
      setSelectedProfiles(
        selectedProfiles.filter((profile) => profile.id !== id),
      );
      if (onRemove) onRemove(id);
    },
    [selectedProfiles, onRemove],
  );

  const isBlocked = useCallback(
    (id) => {
      return (
        blockedIds.findIndex((b) => b === id) !== -1 ||
        selectedProfiles.findIndex((p) => id === p.id) !== -1
      );
    },
    [blockedIds, selectedProfiles],
  );

  useEffect(() => {
    document.addEventListener('keydown', handleEscKey, false);
    document.addEventListener('mousedown', handleDocumentMousedown);
    return () => {
      document.removeEventListener('keydown', handleEscKey, false);
      document.removeEventListener('mousedown', handleDocumentMousedown);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div {...rest}>
      <form aria-label="search-form">
        <DebouncedInput
          inputMode="text"
          value={query}
          className="input debounced-input"
          placeholder="Profile Search ..."
          aria-label="search-profiles"
          ref={searchRef}
          onChange={handleInputChange}
          onFocus={handleInputFocus}
          isLoading={isSearching}
        />
      </form>

      {isError && <Error />}

      {!dropdownHidden && profiles && (
        <div
          className="search-profile-dropdown"
          ref={dropdownRef}
          aria-label="search-dropdown"
        >
          {profiles.map((profile: User) => (
            <ProfileListItem
              profile={profile}
              key={profile.id}
              handleClick={handleItemSelect}
              blocked={isBlocked(profile.id)}
            />
          ))}
        </div>
      )}

      {selectedProfiles && (
        <div aria-label="selected-profiles-list">
          {selectedProfiles.map((profile: User) => (
            <ProfileListItem
              key={profile.id}
              profile={profile}
              chosen
              handleItemRemove={handleItemUnselect}
            />
          ))}
        </div>
      )}
    </div>
  );
}

export default React.memo(SearchProfile);
