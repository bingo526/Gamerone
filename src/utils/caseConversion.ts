export const capFLetter = (value: string) => {
  return value[0].toUpperCase() + value.slice(1);
};

/**
 * Rewrite string in camelCase
 */
const toCamel = (s: string): string => {
  return s.replace(/([-_][a-z])/gi, ($1) => {
    return $1.toUpperCase().replace('-', '').replace('_', '');
  });
};

/**
 * Rwrite string in snake_case
 */
const toSnake = (s: string): string => {
  return s
    .replace(/\W+/g, ' ')
    .split(/ |\B(?=[A-Z])/)
    .map((word) => word.toLowerCase())
    .join('_');
};

/**
 * Check if param is type of array
 */
const isArray = function (a: any) {
  return Array.isArray(a);
};

/**
 * Check if param is type of object
 */
const isObject = function (o: any) {
  return o === Object(o) && !isArray(o) && typeof o !== 'function';
};

/**
 * Convert object keys to camelCase
 */
export const keysToCamel = function (o: any) {
  if (isObject(o)) {
    const n: { [key: string]: any } = {};

    Object.keys(o).forEach((k) => {
      n[toCamel(k)] = keysToCamel(o[k]);
    });

    return n;
  } else if (isArray(o)) {
    return o.map((i: any) => {
      return keysToCamel(i);
    });
  }

  return o;
};

/**
 * Convert object keys to snake_case
 */
export const keysToSnake = function (o: any) {
  if (isObject(o)) {
    const n: { [key: string]: any } = {};

    Object.keys(o).forEach((k) => {
      n[toSnake(k)] = keysToSnake(o[k]);
    });

    return n;
  } else if (isArray(o)) {
    return o.map((i: any) => {
      return keysToSnake(i);
    });
  }

  return o;
};

export const truncateString = (str: string, num: number) => {
  // If the length of str is less than or equal to num
  // just return str--don't truncate it.
  if (str.length <= num) {
    return str;
  }
  // Return str truncated with '...' concatenated to the end of str.
  return str.slice(0, num) + '...';
};
