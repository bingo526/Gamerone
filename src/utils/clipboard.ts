/**
 * Copy text to clipboard
 * @param value text to copy
 */
export default function copyToClipboard(value: string) {
  const textArea = document.createElement('textarea');
  textArea.value = value;
  document.body.appendChild(textArea);
  textArea.select();
  document.execCommand('Copy');
  textArea.remove();
}
