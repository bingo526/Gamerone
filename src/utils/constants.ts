export const FETCH_DEFAULT_OPTIONS = {
  supportHeaderParams: true,
  headers: {
    'Content-Type': 'application/json; charset=UTF-8',
  },
};

export const API_HOST = process.env.REACT_APP_API_ENDPOINT
  ? process.env.REACT_APP_API_ENDPOINT
  : 'API_ENV_NOT_DEFINED';

export const CDN_URL = process.env.REACT_APP_CDN_URL
  ? process.env.REACT_APP_CDN_URL
  : 'CDN_ENV_NOT_DEFINED';

export const CORS_PROXY_URL = process.env.REACT_APP_CORS_PROXY_URL
  ? process.env.REACT_APP_CORS_PROXY_URL
  : 'CORS_PROXY_ENV_NOT_DEFINED';

export const SHARE_FACEBOOK_APP_ID = process.env.REACT_APP_FACEBOOK_APP_ID;

// Google Tag Manager
export const GTM_ID = process.env.REACT_APP_GTM_ID
  ? process.env.REACT_APP_GTM_ID
  : 'GTM_ID_NOT_DEFINED';
export const GTM_AUTH = process.env.REACT_APP_GTM_AUTH
  ? process.env.REACT_APP_GTM_AUTH
  : 'GTM_AUTH_NOT_DEFINED';
export const GTM_PREVIEW = process.env.REACT_APP_GTM_PREVIEW
  ? process.env.REACT_APP_GTM_PREVIEW
  : 'GTM_PREVIEW_NOT_DEFINED';

export const SITE_URL =
  process.env.NODE_ENV === 'production'
    ? 'https://g1.gg'
    : 'https://stage.g1.gg';

export const SITE_NAME = 'Gamer One';
export const SUPPORT_EMAIL = 'support@g1.gg';

// eslint-disable-next-line no-useless-escape
export const CHECK_EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

export const CHECK_PASSWORD_REGEX = /^(?=[^A-Z]*[A-Z])(?=(?:[^a-z]*[a-z]))(?=(?:[^0-9]*[0-9]))(?=(?:[^!?@*#&$._-]*[!?@*#&$._-])).{8,}$/;

export const URL_CHECK_REGEX = new RegExp(
  // eslint-disable-next-line no-useless-escape
  /^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$/,
  'gm',
);

export const MULTI_URL_CHECK_REGEX = new RegExp(
  // eslint-disable-next-line no-useless-escape
  /(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+/,
  'gm',
);
export const URL_PREFIX_CHECK_REGEX = /^(?:http(s)?:\/\/)/;

// Image placeholders
export const AVATAR_PLACEHOLDER = 'assets/placeholder/avatar.svg';
export const COVER_PLACEHOLDER = 'assets/placeholder/cover-image-dark.jpg';
export const GAME_PLACEHOLDER = 'assets/placeholder/game.svg';
export const GEAR_PLACEHOLDER = 'assets/placeholder/gear.svg';
export const POST_PLACEHOLDER = 'assets/placeholder/post.svg';

export const AVATAR_WIDTH = 512; //px
export const AVATAR_HEIGHT = 512; //px

export const COVER_WIDTH = 1600; //px
export const COVER_HEIGHT = 400; //px

export const POST_WIDTH = 640; //px
export const POST_HEIGHT = 360; //px
export const POST_IMAGE_LIMIT = 6;

// Modal default size
export const MODAL_WIDTH = 800; //px
export const MODAL_HEIGHT = 200; //px

export const DEFAULT_USER_VISIBILITY = {
  sponsors: true,
  socialNetworks: true,
  nowPlaying: true,
  gear: true,
  achievements: true,
  store: true,
  squad: true,
  schedule: false,
  timeline: true,
};

export const DEFAULT_CLUB_VISIBILITY = {
  sponsors: true,
  socialNetworks: true,
  nowPlaying: false,
  gear: false,
  achievements: false,
  store: true,
  squad: false,
  schedule: false,
  timeline: true,
};

export const SOCIAL_NETWORK_ICON_LIST = {
  discord: 'icon-video-games-discord',
  facebook: 'icon-social-media-facebook',
  instagram: 'icon-social-instagram',
  reddit: 'icon-social-media-reddit',
  snapchat: 'icon-social-media-snapchat',
  twitch: 'icon-video-game-logo-twitch',
  twitter: 'icon-social-media-twitter',
  weibo: 'icon-feed-sina-weibo',
  vk: 'icon-social-media-vk',
  youtube: 'icon-social-video-youtube-clip',
  linkedin: 'icon-professional-network-linkedin',
};

export const ACHIEVEMENT_ICON_LIST = {
  'up-vote': 'icon-award-trophy-star',
  comment: 'icon-award-trophy-star',
  experience: 'icon-video-game-sword',
  follow: 'icon-award-trophy-star',
  follower: 'icon-video-game-controller-team',
  post: 'icon-quill-write',
  // eslint-disable-next-line @typescript-eslint/camelcase
  profile_complete: 'icon-award-trophy-star',
  signup: 'icon-award-trophy-star',
  game: 'icon-video-game-pacman',
  // eslint-disable-next-line @typescript-eslint/camelcase
  currently_playing: 'icon-award-trophy-star',
  sponsor: 'icon-award-trophy-star',
  socialNetwork: 'icon-award-trophy-star',
  store: 'icon-award-trophy-star',
  gear: 'icon-keyboard',
};

export const CARDS = [
  { id: 1, name: 'sponsors', label: 'Sponsors' },
  { id: 2, name: 'socialNetworks', label: 'Social Networks' },
  { id: 3, name: 'nowPlaying', label: 'Now Playing' },
  { id: 4, name: 'gear', label: 'Gear' },
  { id: 5, name: 'achievements', label: 'Achievements' },
  { id: 6, name: 'store', label: 'Store' },
  { id: 7, name: 'squad', label: 'Squad' },
  // { id: 8, name: 'schedule', label: 'Schedule' },
  // { id: 9, name: 'teams', label: 'Teams' },
];

// Now Playing
export const STREAM_MEDIAS = ['facebook', 'twitch', 'youtube'];

export const G1_THEMES = {
  Default: 'theme--1',
  Funky: 'theme--2',
  Techie: 'theme--3',
  Soft: 'theme--4',
  Elegant: 'theme--5',
} as const;
