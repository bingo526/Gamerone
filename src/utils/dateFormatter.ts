import dayjs from 'dayjs';

export const formatDateDayMonth = (date: number | Date) => {
  return new Intl.DateTimeFormat('en-US', {
    day: '2-digit',
    month: 'short',
  }).format(date);
};

export const formatPureDateDayMonth = (date: number | string | Date) => {
  return dayjs(date).format('MMM DD');
};

export const formatPureDateMonthYear = (date: number | string | Date) => {
  return dayjs(date).format('MMM YYYY');
};

export const durationBetween = (
  startDate: number | string | Date,
  endDate: number | string | Date,
  unit: 'day' | 'month' | 'year',
) => {
  return dayjs(endDate).diff(startDate, unit, true);
};

export const formatDateDayMonthYear = (date: number | Date) => {
  return new Intl.DateTimeFormat('en-US', {
    day: '2-digit',
    month: 'short',
    year: 'numeric',
  }).format(date);
};
