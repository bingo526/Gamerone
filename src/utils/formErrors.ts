export const ERROR_MSG_USERNAME_DUPLICATE = 'Username already exists.';
export const ERROR_MSG_EMAIL_PATTERN = 'Email format incorrect.';
export const ERROR_MSG_EMAIL_DUPLICATE = 'Email already exists.';
export const ERROR_MSG_MIN_LENGTH_2 = 'Must be at least 2 characters long.';
export const ERROR_MSG_MIN_LENGTH_3 = 'Must be at least 3 characters long.';
export const ERROR_MSG_MIN_LENGTH_8 = 'Must be at least 8 characters long.';
export const ERROR_MSG_DUPLICATE = 'Duplicated';
export const ERROR_MSG_PASSWORD_PATTERN =
  'Password must contain at least one uppercase, one lowercase, one digit and one special letter.';
export const ERROR_MSG_PASSWORD_MATCH = 'Passwords do not match.';
export const ERROR_MSG_URL_PATTERN =
  'URL format is incorrect. Must start with https:// or yourdomain.';
