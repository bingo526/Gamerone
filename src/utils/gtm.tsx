import TagManager from 'react-gtm-module';

/**
 * Event - Add custom tracking event.
 * This is for tracking when event clicked something like that they'll send the data to GA
 */
export const GTMDataLayer = (event: string, data: any) => {
  TagManager.dataLayer({
    dataLayerName: 'GamerOne',
    dataLayer: {
      event: event,
      data: data,
    },
  });
};

export default GTMDataLayer;
