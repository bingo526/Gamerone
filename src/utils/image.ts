import { CDN_URL, CORS_PROXY_URL } from './constants';

/**
 * Convert url to File object
 *
 * @param url string
 */
export const getFileFromUrl = async (url: string) => {
  let nLink = url;
  if (url.startsWith('user_content')) {
    nLink = CDN_URL + '/512/' + url;
  } else if (!url.startsWith('blob:')) {
    nLink = `${CORS_PROXY_URL}/${url}`;
  }

  const blob = await fetch(nLink).then((r: any) => r.blob());

  const file = new File([blob], 'f.png', {
    type: blob.type,
    lastModified: Date.now(),
  });

  return file;
};

/**
 * Crop image to a specific aspect ratio
 *
 * @url - Source of the image to use
 * @aspectRatio - The aspect ratio to apply
 */
export const crop = (url: string, aspectRatio: number) => {
  return new Promise<HTMLCanvasElement>((resolve) => {
    // this image will hold our source image data
    const inputImage = new Image();

    // we want to wait for our image to load
    inputImage.onload = () => {
      // let's store the width and height of our image
      const inputWidth = inputImage.naturalWidth;
      const inputHeight = inputImage.naturalHeight;

      // get the aspect ratio of the input image
      const inputImageAspectRatio = inputWidth / inputHeight;

      // if it's bigger than our target aspect ratio
      let outputWidth = inputWidth;
      let outputHeight = inputHeight;
      if (inputImageAspectRatio > aspectRatio) {
        outputWidth = inputHeight * aspectRatio;
      } else if (inputImageAspectRatio < aspectRatio) {
        outputHeight = inputWidth / aspectRatio;
      }

      // calculate the position to draw the image at
      const outputX = (outputWidth - inputWidth) * 0.5;
      const outputY = (outputHeight - inputHeight) * 0.5;

      // create a canvas that will present the output image
      const outputImage = document.createElement('canvas');

      // set it to the same size as the image
      outputImage.width = outputWidth;
      outputImage.height = outputHeight;

      // draw our image at position 0, 0 on the canvas
      const ctx = outputImage.getContext('2d');
      if (ctx) ctx.drawImage(inputImage, outputX, outputY);
      resolve(outputImage);
    };

    // start loading our image
    inputImage.src = url;
  });
};

export const cropImage = async (url: string, aspectRatio: number) => {
  const canvas = await crop(url, aspectRatio);
  const response = await fetch(canvas.toDataURL());

  return window.URL.createObjectURL(response.blob());
};
