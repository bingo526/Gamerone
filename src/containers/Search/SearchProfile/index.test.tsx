import React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter, Route } from 'react-router-dom';
import SearchProfile from './index';
import { AVATAR_PLACEHOLDER } from 'utils/constants';

const defaultProfiles = [
  {
    id: 11,
    username: 'pokimane',
    email: 'reserved+pokimane@gamerone.gg',
    firstName: 'Pokimane',
    lastName: 'User',
    bio: null,
    avatar: AVATAR_PLACEHOLDER,
    banner: null,
    playCount: 0,
    gameCount: 0,
    badgeCount: 0,
    followCount: 0,
    followerCount: 0,
    birthDate: null,
  },
  {
    id: 13,
    username: 'drlupo',
    email: 'reserved+drlupo@gamerone.gg',
    firstName: 'Drlupo',
    lastName: 'User',
    bio: null,
    avatar: AVATAR_PLACEHOLDER,
    banner: null,
    playCount: 0,
    gameCount: 0,
    badgeCount: 0,
    followCount: 0,
    followerCount: 0,
    birthDate: null,
  },
  {
    id: 16,
    username: 'nickmercs',
    email: 'reserved+nickmercs@gamerone.gg',
    firstName: 'Nickmercs',
    lastName: 'User',
    bio: null,
    avatar: AVATAR_PLACEHOLDER,
    banner: null,
    playCount: 0,
    gameCount: 0,
    badgeCount: 0,
    followCount: 0,
    followerCount: 0,
    birthDate: null,
  },
];

const searchProfile = async (
  {
    multiple = false,
    navigate = false,
    clearInput = false,
    initLoad = false,
  } = {},
  additions = null,
) => {
  const jestFetch = jest
    .spyOn(window, 'fetch')
    .mockImplementationOnce(async () => {
      return Promise.resolve({
        status: 200,
        json: () => defaultProfiles,
      });
    });

  const utils = render(
    <MemoryRouter>
      <SearchProfile
        multiple={multiple}
        navigate={navigate}
        clearInput={clearInput}
        initLoad={initLoad}
      />
      {additions}
    </MemoryRouter>,
  );

  act(() => {
    fireEvent.change(utils.getByLabelText(/search-profiles/), {
      target: { value: 'profile' },
    });
  });

  await waitForElement(() => utils.getByLabelText(/search-dropdown/));
  await waitForElement(() => utils.getAllByLabelText(/profile-list-item/));

  jestFetch.mockRestore();
  return { ...utils, jestFetch };
};

describe('basic functionality', () => {
  test('searches profiles, shows the dropdown and hides it on outside click', async () => {
    const {
      asFragment,
      getByLabelText,
      queryByLabelText,
    } = await searchProfile();
    expect(asFragment()).toMatchSnapshot();

    act(() => {
      fireEvent.mouseDown(getByLabelText(/search-form/));
    });

    expect(queryByLabelText(/search-dropdown/)).toBe(null);
  });

  test('input stays the focused after search', async () => {
    const { getByLabelText } = await searchProfile();

    const input = getByLabelText(/search-profiles/);
    expect(document.activeElement).toBe(input);
  });
});

describe('option: multiple', () => {
  test('adds item to the list on click and not hide the dropdown', async () => {
    const { getByLabelText } = await searchProfile({ multiple: true });

    const dropdown = getByLabelText(/search-dropdown/);
    expect(dropdown).toBeInTheDocument();

    const firstProfile = dropdown.querySelectorAll('.profile-list-item')[0];
    act(() => fireEvent.click(firstProfile));

    const selectedList = getByLabelText(/selected-profiles-list/);
    expect(selectedList).toBeInTheDocument();

    const selectedProfileName = selectedList.querySelectorAll(
      '.profile-name',
    )[0];

    // ensure that same named item is added
    expect(firstProfile.querySelector('.profile-name').innerHTML).toBe(
      selectedProfileName.innerHTML,
    );

    expect(dropdown).toBeInTheDocument();
  });
});

describe('option: navigate', () => {
  test('navigates to the profile page on item click', async () => {
    let capturedlocation;
    const { getByLabelText } = await searchProfile(
      { navigate: true },
      <Route
        path="*"
        render={({ history, location }) => {
          capturedlocation = location;
          return null;
        }}
      />,
    );

    const dropdown = getByLabelText(/search-dropdown/);
    expect(dropdown).toBeInTheDocument();

    const firstProfile = dropdown.querySelectorAll('.profile-list-item')[0];
    const profileName = firstProfile
      .querySelector('.profile-name span')
      .innerHTML.substring(1);
    act(() => fireEvent.click(firstProfile));

    expect(capturedlocation.pathname).toBe(`/${profileName}`);
  });
});

describe('option: clearInput', () => {
  test('clears the input after the item is clicked', async () => {
    const { getByLabelText, asFragment } = await searchProfile({
      clearInput: true,
    });

    const dropdown = getByLabelText(/search-dropdown/);
    expect(dropdown).toBeInTheDocument();

    const firstProfile = dropdown.querySelectorAll('.profile-list-item')[0];
    act(() => fireEvent.click(firstProfile));

    expect(asFragment()).toMatchSnapshot();
    const input = getByLabelText(/search-profiles/);

    expect(input.value).toBe('');
  });
});

describe('option: initLoad', () => {
  test('should fetch data and show the dropdown initially', async () => {
    const jestFetch = jest
      .spyOn(window, 'fetch')
      .mockImplementationOnce(async () => {
        return Promise.resolve({
          status: 200,
          json: () => defaultProfiles,
        });
      });

    const { getAllByLabelText, getByLabelText, asFragment } = render(
      <MemoryRouter>
        <SearchProfile initLoad={true} />
      </MemoryRouter>,
    );

    await waitForElement(() => getByLabelText(/search-dropdown/));
    await waitForElement(() => getAllByLabelText(/profile-list-item/));

    expect(asFragment()).toMatchSnapshot();

    expect(jestFetch).toBeCalled();
    expect(jestFetch).toHaveBeenCalledTimes(1);

    const items = getAllByLabelText(/profile-list-item/);
    expect(items.length).toBe(defaultProfiles.length);

    jestFetch.mockRestore();
  });

  test('should not fetch data and not show the dropdown on load', async () => {
    const jestFetch = jest
      .spyOn(window, 'fetch')
      .mockImplementationOnce(async () => {
        return Promise.resolve({
          status: 200,
          json: () => defaultProfiles,
        });
      });

    const { queryAllByLabelText, getByLabelText, asFragment } = render(
      <MemoryRouter>
        <SearchProfile initLoad={false} />
      </MemoryRouter>,
    );

    await waitForElement(() => getByLabelText(/search-dropdown/));

    const items = queryAllByLabelText(/profile-list-item/);
    expect(items.length).toBe(0);

    expect(asFragment()).toMatchSnapshot();
    expect(jestFetch).toHaveBeenCalledTimes(0);

    jestFetch.mockRestore();
  });
});
