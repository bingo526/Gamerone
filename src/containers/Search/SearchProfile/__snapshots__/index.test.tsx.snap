// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic functionality searches profiles, shows the dropdown and hides it on outside click 1`] = `
<DocumentFragment>
  <div>
    <form
      aria-label="search-form"
      class="form-inline"
    >
      <div
        class="position-relative d-block w-100"
      >
        <input
          aria-label="search-profiles"
          class="form-control w-100 debounced-input"
          inputmode="text"
          placeholder="Search Profile"
          value="profile"
        />
      </div>
    </form>
    <div
      aria-label="search-dropdown"
      class="search-profile-dropdown"
    >
      <div
        aria-label="profile-list-item"
        class="d-flex profile-list-item"
      >
        <div
          class="p-2 profile-avatar"
        >
          <img
            alt="pokimane"
            class="avatar lg border border-body border-width-2"
            id="avatar"
            src="https://cdn.g1.gg/128/assets/placeholder/avatar.svg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 profile-name text-wrap text-secondary"
        >
          <h6>
            Pokimane User
          </h6>
          <span>
            @pokimane
          </span>
        </div>
      </div>
      <div
        aria-label="profile-list-item"
        class="d-flex profile-list-item"
      >
        <div
          class="p-2 profile-avatar"
        >
          <img
            alt="drlupo"
            class="avatar lg border border-body border-width-2"
            id="avatar"
            src="https://cdn.g1.gg/128/assets/placeholder/avatar.svg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 profile-name text-wrap text-secondary"
        >
          <h6>
            Drlupo User
          </h6>
          <span>
            @drlupo
          </span>
        </div>
      </div>
      <div
        aria-label="profile-list-item"
        class="d-flex profile-list-item"
      >
        <div
          class="p-2 profile-avatar"
        >
          <img
            alt="nickmercs"
            class="avatar lg border border-body border-width-2"
            id="avatar"
            src="https://cdn.g1.gg/128/assets/placeholder/avatar.svg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 profile-name text-wrap text-secondary"
        >
          <h6>
            Nickmercs User
          </h6>
          <span>
            @nickmercs
          </span>
        </div>
      </div>
    </div>
    <div
      aria-label="selected-profiles-list"
      class="d-flex flex-wrap"
    />
  </div>
</DocumentFragment>
`;

exports[`option: clearInput clears the input after the item is clicked 1`] = `
<DocumentFragment>
  <div>
    <form
      aria-label="search-form"
      class="form-inline"
    >
      <div
        class="position-relative d-block w-100"
      >
        <input
          aria-label="search-profiles"
          class="form-control w-100 debounced-input"
          inputmode="text"
          placeholder="Search Profile"
          value=""
        />
      </div>
    </form>
    <div
      aria-label="selected-profiles-list"
      class="d-flex flex-wrap"
    >
      <div
        aria-label="profile-list-item"
        class="d-flex profile-list-item col-3 chosen"
      >
        <div
          class="p-2 profile-avatar"
        >
          <img
            alt="pokimane"
            class="avatar lg border border-body border-width-2"
            id="avatar"
            src="https://cdn.g1.gg/128/assets/placeholder/avatar.svg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 profile-name text-wrap text-secondary"
        >
          <h6>
            Pokimane User
          </h6>
          <span>
            @pokimane
          </span>
        </div>
        <span
          class="item-remove"
        >
          <i
            class="far fa-times-circle"
          />
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`option: initLoad should fetch data and show the dropdown initially 1`] = `
<DocumentFragment>
  <div>
    <form
      aria-label="search-form"
      class="form-inline"
    >
      <div
        class="position-relative d-block w-100"
      >
        <input
          aria-label="search-profiles"
          class="form-control w-100 debounced-input"
          inputmode="text"
          placeholder="Search Profile"
          value=""
        />
      </div>
    </form>
    <div
      aria-label="search-dropdown"
      class="search-profile-dropdown"
    >
      <div
        aria-label="profile-list-item"
        class="d-flex profile-list-item"
      >
        <div
          class="p-2 profile-avatar"
        >
          <img
            alt="pokimane"
            class="avatar lg border border-body border-width-2"
            id="avatar"
            src="https://cdn.g1.gg/128/assets/placeholder/avatar.svg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 profile-name text-wrap text-secondary"
        >
          <h6>
            Pokimane User
          </h6>
          <span>
            @pokimane
          </span>
        </div>
      </div>
      <div
        aria-label="profile-list-item"
        class="d-flex profile-list-item"
      >
        <div
          class="p-2 profile-avatar"
        >
          <img
            alt="drlupo"
            class="avatar lg border border-body border-width-2"
            id="avatar"
            src="https://cdn.g1.gg/128/assets/placeholder/avatar.svg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 profile-name text-wrap text-secondary"
        >
          <h6>
            Drlupo User
          </h6>
          <span>
            @drlupo
          </span>
        </div>
      </div>
      <div
        aria-label="profile-list-item"
        class="d-flex profile-list-item"
      >
        <div
          class="p-2 profile-avatar"
        >
          <img
            alt="nickmercs"
            class="avatar lg border border-body border-width-2"
            id="avatar"
            src="https://cdn.g1.gg/128/assets/placeholder/avatar.svg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 profile-name text-wrap text-secondary"
        >
          <h6>
            Nickmercs User
          </h6>
          <span>
            @nickmercs
          </span>
        </div>
      </div>
    </div>
    <div
      aria-label="selected-profiles-list"
      class="d-flex flex-wrap"
    />
  </div>
</DocumentFragment>
`;

exports[`option: initLoad should not fetch data and not show the dropdown on load 1`] = `
<DocumentFragment>
  <div>
    <form
      aria-label="search-form"
      class="form-inline"
    >
      <div
        class="position-relative d-block w-100"
      >
        <input
          aria-label="search-profiles"
          class="form-control w-100 debounced-input"
          inputmode="text"
          placeholder="Search Profile"
          value=""
        />
      </div>
    </form>
    <div
      aria-label="search-dropdown"
      class="search-profile-dropdown"
    />
    <div
      aria-label="selected-profiles-list"
      class="d-flex flex-wrap"
    />
  </div>
</DocumentFragment>
`;
